File: clients/go/pool.go
// Package lightkv provides a connection pool for LightKV client SDK.
// Wraps Client with a pool of reusable TCP connections.
package lightkv

import (
	"errors"
	"sync"
	"time"
)

// Pool is a connection pool for LightKV Client.
//
// Usage:
//
//	pool := NewPool("127.0.0.1:6379", PoolOptions{PoolSize: 10})
//	c, err := pool.Acquire()
//	if err != nil { panic(err) }
//	defer pool.Release(c)
//	c.Set("key", "value")
//	pool.Close()
type Pool struct {
	addr      string
	timeout   time.Duration
	poolSize  int
	retry     int
	idleTTL   time.Duration
	healthCheck time.Duration
	pingTimeout time.Duration

	mu        sync.Mutex
	idle      []*poolEntry
	closed    bool
	health    *poolHealth
}

type poolEntry struct {
	client      *Client
	lastUsedAt time.Time
	lastHealthAt time.Time
}

// PoolOptions configures a Pool.
type PoolOptions struct {
	Timeout    time.Duration // connection timeout (default 5s)
	PoolSize   int           // max cached connections (default 10)
	Retry      int           // retry attempts on connection failure (default 3)
	IdleTTL    time.Duration // idle connection TTL (default 5m)
	HealthCheck time.Duration // health check interval (default 1m); 0 = disable
	PingTimeout time.Duration // PING timeout for health check (default 1s)
}

// health state for background checker
type poolHealth struct {
	stop chan struct{}
	wg   sync.WaitGroup
}

// NewPool creates a connection pool for the given LightKV server address.
func NewPool(addr string, opts PoolOptions) *Pool {
	if addr == "" {
		addr = "127.0.0.1:6379"
	}
	if opts.Timeout == 0 {
		opts.Timeout = 5 * time.Second
	}
	if opts.PoolSize == 0 {
		opts.PoolSize = 10
	}
	if opts.IdleTTL == 0 {
		opts.IdleTTL = 5 * time.Minute
	}
	if opts.HealthCheck == 0 {
		opts.HealthCheck = 1 * time.Minute
	}
	if opts.PingTimeout == 0 {
		opts.PingTimeout = 1 * time.Second
	}
	p := &Pool{
		addr:     addr,
		timeout:  opts.Timeout,
		poolSize: opts.PoolSize,
		retry:    opts.Retry,
		idleTTL:  opts.IdleTTL,
		healthCheck: opts.HealthCheck,
		pingTimeout: opts.PingTimeout,
	}
	if opts.HealthCheck > 0 {
		p.health = &poolHealth{stop: make(chan struct{})}
		p.health.wg.Add(1)
		go p.healthLoop()
	}
	return p
}

// healthLoop 后台健康检查：定期 PING idle 连接，剔除坏的
func (p *Pool) healthLoop() {
	defer p.health.wg.Done()
	ticker := time.NewTicker(p.healthCheck)
	defer ticker.Stop()
	for {
		select {
		case <-p.health.stop:
			return
		case <-ticker.C:
			p.doHealthCheck()
		}
	}
}

func (p *Pool) doHealthCheck() {
	now := time.Now()
	var alive []*poolEntry
	p.mu.Lock()
	for _, e := range p.idle {
		if now.Sub(e.lastUsedAt) > p.idleTTL {
			e.client.Close()
			continue
		}
		if now.Sub(e.lastHealthAt) < p.healthCheck {
			alive = append(alive, e)
			continue
		}
		if p.pingAlive(e.client) {
			e.lastHealthAt = now
			alive = append(alive, e)
		} else {
			e.client.Close()
		}
	}
	p.idle = alive
	p.mu.Unlock()
}

func (p *Pool) pingAlive(c *Client) bool {
	// PING with timeout; if client has no Ping method, treat as alive
	// Use a short deadline via a stub call (Client doesn't expose PING yet)
	// For safety, return true (alive) — real PING added when Client exposes it
	return true
}

// newClient creates a Client with exponential backoff retry.
func (p *Pool) newClient() (*Client, error) {
	var lastErr error
	for attempt := 0; attempt <= p.retry; attempt++ {
		c := NewClient(p.addr, p.timeout)
		if err := c.Connect(); err == nil {
			return c, nil
		} else {
			lastErr = err
			if attempt < p.retry {
				time.Sleep(time.Duration(100<<(attempt)) * time.Millisecond)
			}
		}
	}
	return nil, errors.New("failed to connect after retries: " + lastErr.Error())
}

// Acquire takes a client from the pool, or creates a new one if pool is empty.
func (p *Pool) Acquire() (*Client, error) {
	if p.closed {
		return nil, errors.New("pool is closed")
	}
	for {
		p.mu.Lock()
		if len(p.idle) == 0 {
			p.mu.Unlock()
			return p.newClient()
		}
		entry := p.idle[len(p.idle)-1]
		p.idle = p.idle[:len(p.idle)-1]
		p.mu.Unlock()
		// Stale connection → discard and retry
		if time.Since(entry.lastUsedAt) > p.idleTTL {
			entry.client.Close()
			continue
		}
		return entry.client, nil
	}
}

// Release returns a client to the pool for reuse, or closes it if pool is full.
func (p *Pool) Release(c *Client) {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.closed || len(p.idle) >= p.poolSize {
		c.Close()
		return
	}
	now := time.Now()
	p.idle = append(p.idle, &poolEntry{client: c, lastUsedAt: now, lastHealthAt: now})
}

// Close closes all idle connections and marks pool as closed.
func (p *Pool) Close() {
	p.mu.Lock()
	p.closed = true
	if p.health != nil {
		close(p.health.stop)
	}
	for _, e := range p.idle {
		e.client.Close()
	}
	p.idle = nil
	p.mu.Unlock()
	if p.health != nil {
		p.health.wg.Wait()
		p.health = nil
	}
}

File: clients/nodejs/src/pool.js
'use strict';

const LightKVClient = require('./client.js');

/**
 * LightKV Connection Pool for Node.js SDK.
 * Wraps LightKVClient with a pool of reusable TCP connections.
 *
 * @example
 * const { LightKVPool } = require('./pool.js');
 * const pool = new LightKVPool({ host: '127.0.0.1', port: 6379, poolSize: 10 });
 * const client = await pool.acquire();
 * try {
 *   await client.set('key', 'value');
 *   console.log(await client.get('key'));
 * } finally {
 *   await pool.release(client);
 * }
 * await pool.close();
 */
class LightKVPool {
  /**
   * @param {Object} [options]
   * @param {string} [options.host='127.0.0.1']
   * @param {number} [options.port=6379]
   * @param {number} [options.timeout=5000] - Connection timeout in ms
   * @param {number} [options.poolSize=10] - Maximum cached connections
   * @param {number} [options.retry=3] - Retry attempts on connection failure
   * @param {number} [options.idleTimeout=300000] - Idle timeout in ms
   */
  constructor(options = {}) {
    this.host = options.host || '127.0.0.1';
    this.port = options.port || 6379;
    this.timeout = options.timeout || 5000;
    this.poolSize = options.poolSize || 10;
    this.retry = options.retry || 3;
    this.idleTimeout = options.idleTimeout || 300000;
    this.healthCheckInterval = options.healthCheckInterval || 60000;
    this.pingTimeout = options.pingTimeout || 1000;

    this._idle = []; // [{ client, lastUsedAt, lastHealthAt }
    this._closed = false;
    this._healthTimer = null;
    if (this.healthCheckInterval > 0) {
      this._healthTimer = setInterval(() => this._healthCheck(), this.healthCheckInterval);
      if (this._healthTimer.unref) this._healthTimer.unref();
    }
  }

  // 后台健康检查：定期 PING idle 连接，剔除坏的
  async _healthCheck() {
    if (this._closed) return;
    const now = Date.now();
    const alive = [];
    for (const entry of this._idle) {
      const { client, lastUsedAt, lastHealthAt } = entry;
      if (now - lastUsedAt > this.idleTimeout) {
        try { await client.disconnect(); } catch (e) {}
        continue;
      }
      if (now - lastHealthAt < this.healthCheckInterval) {
        alive.push(entry);
        continue;
      }
      if (await this._pingAlive(client)) {
        entry.lastHealthAt = now;
        alive.push(entry);
      } else {
        try { await client.disconnect(); } catch (e) {}
      }
    }
    this._idle = alive;
  }

  async _pingAlive(client) {
    try {
      if (typeof client.ping === 'function') {
        return await client.ping();
      }
      return true;
    } catch (e) {
      return false;
    }
  }

  async _newClient() {
    let lastErr = null;
    for (let attempt = 0; attempt <= this.retry; attempt++) {
      const c = new LightKVClient({ host: this.host, port: this.port, timeout: this.timeout });
      try {
        await c.connect();
        return c;
      } catch (e) {
        lastErr = e;
        if (attempt < this.retry) {
          await new Promise(r => setTimeout(r, 100 * (2 ** attempt)));
        }
      }
    }
    throw new Error(`failed to connect after ${this.retry + 1} attempts: ${lastErr && lastErr.message}`);
  }

  /**
   * Acquire a client from the pool (or create a new one).
   * @returns {Promise<LightKVClient>}
   */
  async acquire() {
    if (this._closed) throw new Error('pool is closed');
    while (this._idle.length > 0) {
      const { client, lastUsedAt } = this._idle.pop();
      // Stale connection → discard and retry
      if (Date.now() - lastUsedAt > this.idleTimeout) {
        try { await client.disconnect(); } catch (e) {}
        continue;
      }
      return client;
    }
    return this._newClient();
  }

  /**
   * Release a client back to the pool for reuse.
   * @param {LightKVClient} client
   */
  async release(client) {
    if (this._closed || this._idle.length >= this.poolSize) {
      try { await client.disconnect(); } catch (e) {}
      return;
    }
    this._idle.push({ client, lastUsedAt: Date.now(), lastHealthAt: Date.now() });
  }

  /**
   * Close all idle connections and mark pool as closed.
   */
  async close() {
    this._closed = true;
    if (this._healthTimer) {
      clearInterval(this._healthTimer);
      this._healthTimer = null;
    }
    for (const { client } of this._idle) {
      try { await client.disconnect(); } catch (e) {}
    }
    this._idle = [];
  }
}

module.exports = { LightKVPool };
